py4cast/models/vision/unetrpp.py

```diff
@@ -266,21 +291,38 @@ def forward(self, x):
         q_shared = torch.nn.functional.normalize(q_shared, dim=-1)
         k_shared = torch.nn.functional.normalize(k_shared, dim=-1)
 
-        attn_CA = (q_shared @ k_shared.transpose(-2, -1)) * self.temperature
-        attn_CA = attn_CA.softmax(dim=-1)
-        attn_CA = self.attn_drop(attn_CA)
-        x_CA = (attn_CA @ v_CA).permute(0, 3, 1, 2).reshape(B, N, C)
-
-        attn_SA = (
-            q_shared.permute(0, 1, 3, 2) @ k_shared_projected
-        ) * self.temperature2
-        attn_SA = attn_SA.softmax(dim=-1)
-        attn_SA = self.attn_drop_2(attn_SA)
-        x_SA = (
-            (attn_SA @ v_SA_projected.transpose(-2, -1))
-            .permute(0, 3, 1, 2)
-            .reshape(B, N, C)
-        )
+        if self.use_scaled_dot_product_CA:
+            x_CA = scaled_dot_product_attention(
+                q_shared, k_shared, v_CA, dropout_p=self.attn_drop.p
+            )
+        else:
+            attn_CA = (q_shared @ k_shared.transpose(-2, -1)) * self.temperature
+            attn_CA = attn_CA.softmax(dim=-1)
+            attn_CA = self.attn_drop(attn_CA)
+            x_CA = attn_CA @ v_CA
+
+        x_CA = x_CA.permute(0, 3, 1, 2).reshape(B, N, C)
+
+        if self.use_scaled_dot_product_SA:
+            q_shared_projected = torch.einsum("bhdn,nk->bhdk", q_shared, self.EF)
```

Comment:
The `self.temperature` play no role with sdpa but they are used in the other case : is this a desired behaviour ?
(like, adding self.temperature as the "scale" parameter ?)

Reply:
I tried passing the temperature as the scale argument of sdpa but it doesn"t accept trainable parameters. This is a design inherited from the original paper implementation so yes it is desired in the sense that it is the best compromise i found to use sdpa.

Reply:
What about multipliying `q_shared` / `k_shared` by `torch.rsqrt(self.temperature)` beforehand and keep the scale kwargs to 1.0 ?

Reply:
We could try yes, also compare the results with and without it, and if it doesn't improve the performances, we could probably get rid of it altogether ?

Reply:
ok, later pr